### DISCUSSION TITLE
Date objects not persisted correctly

### DISCUSSION THREAD

**Discussion by mfkrause** [pmndrs/zustand#1720] [General] [answered]
First of all: Not quite sure if this would fit in the Q&A or ideas category as I'm still unsure if I'm simply missing something or this is indeed something that hasn't been thought about so far, so I'll just keep it to general for now.

Using the new `createJSONStorage()`, it doesn't seem to be possible anymore to provide a custom `deserialize()` function. Under the hood, `createJSONStorage()` simply uses a vanilla `JSON.parse()` call to deserialize the string value from the persisted storage.

Unfortunately, this poses a problem when there are Date objects persisted. The native `JSON.parse()` will not deserialize them into Date objects again but will keep them as Strings instead. This creates an inconsistency with the state types and is — obviously — prone to creating further bugs.

I've tried a few things to work around this but haven't succeeded so far, as the `StateStorage` expects only strings in both the `getItem` and `setItem` properties. However, the `JSON.parse()` API has an optional `reviver` parameter. See [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse). If the `createJSONStorage()` API would pass this on as an optional configuration option, one could very possibly work around this and write their own reviver.

I've spun up a little patch which works well for me:
```diff
diff --git a/node_modules/zustand/middleware.js b/node_modules/zustand/middleware.js
index ef6c4b4..bdadcd4 100644
--- a/node_modules/zustand/middleware.js
+++ b/node_modules/zustand/middleware.js
@@ -286,7 +286,7 @@ var combine = function combine(initialState, create) {
   };
 };
 
-function createJSONStorage(getStorage) {
+function createJSONStorage(getStorage, reviver) {
   var storage;
   try {
     storage = getStorage();
@@ -300,7 +300,7 @@ function createJSONStorage(getStorage) {
         if (str === null) {
           return null;
         }
-        return JSON.parse(str);
+        return JSON.parse(str, reviver);
       };
       var str = (_getItem = storage.getItem(name)) != null ? _getItem : null;
       if (str instanceof Promise) {
diff --git a/node_modules/zustand/middleware/persist.d.ts b/node_modules/zustand/middleware/persist.d.ts
index c9fb534..0d6bab1 100644
--- a/node_modules/zustand/middleware/persist.d.ts
+++ b/node_modules/zustand/middleware/persist.d.ts
@@ -13,7 +13,7 @@ export interface PersistStorage<S> {
     setItem: (name: string, value: StorageValue<S>) => void | Promise<void>;
     removeItem: (name: string) => void | Promise<void>;
 }
-export declare function createJSONStorage<S>(getStorage: () => StateStorage): PersistStorage<S> | undefined;
+export declare function createJSONStorage<S>(getStorage: () => StateStorage, reviver?: (key: string, value: any) => any): PersistStorage<S> | undefined;
 export interface PersistOptions<S, PersistedState = S> {
     /** Name of the storage (must be unique) */
     name: string;
```

Used like so in my case:
```typescript
{
  name: 'app-state',
  storage: createJSONStorage(
    () => mmkvStorage,
    (_key, value) => {
      // check if value is a Date ISO string
      if (typeof value === 'string' && /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.\d{3}Z$/.test(value)) {
        return new Date(value);
      }
      return value;
    }
  ),
}
```

However, I'm still unsure if I am missing something obvious here that would already allow me to get those Date objects back with the current implementation of zustand. Otherwise, I'd be glad to put this in a PR...

**Comment by JacobWeisenburger** [comment]
I figured out how to do it, but I don't like this way because it's not very intuitive nor flexible.

```ts
import { z } from 'zod'

const store = createStore(
    persist(
        () => ( { date: new Date( '1970-01-01' ) } ),
        {
            name: 'app-state',
            storage: createJSONStorage(
                () => ( {
                    getItem: key => Deno.readTextFile( `./${ key }.json` ),
                    setItem ( key, value ) {
                        Deno.writeTextFile( `./${ key }.json`, value )
                    },
                    removeItem ( key ) {
                        Deno.remove( `./${ key }.json` )
                    },
                } ),
            ),
            onRehydrateStorage ( stateBeforeRehydrate ) {
                console.log( { stateBeforeRehydrate } )
                return stateAfterRehydrate => {
                    console.log( { stateAfterRehydrate } )

                    const stateString = JSON.stringify( stateAfterRehydrate )
                    const newState = JSON.parse(
                        stateString,
                        ( key, value ) => {
                            const isDateString = z.string().datetime().safeParse( value )
                            if ( isDateString.success ) return new Date( isDateString.data )
                            return value
                        }
                    )
                    store.setState( newState )
                }
            },
        }
    )
)

console.log( { initialState: store.getState() } )

/* waits so Deno.readTextFile has time to finish */
await new Promise( resolve => setTimeout( resolve ) )

console.log( { finalState: store.getState() } )

/* LOG */
// { stateBeforeRehydrate: undefined }
// { initialState: { date: 1970-01-01T00:00:00.000Z } } // date as object
// { stateAfterRehydrate: { date: "2023-03-30T16:09:47.017Z" } } // date as string
// { finalState: { date: 2023-03-30T16:09:47.017Z } } // date as object
```

I feel like allowing the developer to return whatever they want from `getItem` and exactly that is used as the state would be better. This way the developer can decide if they want to use the base `JSON.parse` or with a custom reviver, or something completely different. The developer should be allowed to choose.

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by mfkrause** [reply]
While I didn't test this option, this _should_ work. Feels kinda clunky though. I agree that allowing to return any value from getItem() or — like in my patch above — allowing a custom reviver for the JSON parser is a cleaner option.

**Comment by JacobWeisenburger** [comment]
Here is a way to do what you are asking for without using `createJSONStorage`. Let me know if I have missed something or if you have any questions.

```ts
import { z } from 'zod'

const store = createStore(
    persist(
        () => ( { date: {} } ),
        {
            name: 'app-state',
            storage: {
                getItem: async key => {
                    return JSON.parse(
                        await Deno.readTextFile( `./${ key }.json` ),
                        ( key, value ) => {
                            const isDateString = z.string().datetime().safeParse( value )
                            if ( isDateString.success ) return new Date( isDateString.data )
                            return value
                        }
                    )
                },
                async setItem ( key, value ) {
                    await Deno.writeTextFile(
                        `./${ key }.json`,
                        JSON.stringify( value )
                    )
                },
                async removeItem ( key ) {
                    await Deno.remove( `./${ key }.json` )
                },
            },
        }
    )
)

console.log( { initialState: store.getState() } )
// { initialState: { date: { } } }

/* waits so Deno.readTextFile has time to finish */
await new Promise( resolve => setTimeout( resolve ) )

console.log( { finalState: store.getState() } )
// { finalState: { date: 2023-03-30T16:55:23.975Z } }
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by mfkrause** [reply]
Yep, that's what I've tried, too. While this works fine, it's incompatible with the type implementation of zustand. `getItem()` expects a string or null to be returned:
https://github.com/pmndrs/zustand/blob/e489a63513509b62d33e5580d5541680dbcbffb9/src/middleware/persist.ts#L3-L7

Returning a Date object therefore leads to a TypeScript error. So this is only viable in JavaScript, not TypeScript, unfortunately and the reason for my hacky patch using the `revive` parameter.

**Reply by JacobWeisenburger** [reply]
I ran this in Deno in a typescript file and I didn't get any compile errors. Where do you see the compile errors?

**Reply by JacobWeisenburger** [reply]
`StateStorage` is not used by `PersistOptions.storage`, it uses `PersistStorage`

https://github.com/pmndrs/zustand/blob/142b60025b41ed3407c2abb08afe3f048ea6258b/src/middleware/persist.ts#L92

https://github.com/pmndrs/zustand/blob/142b60025b41ed3407c2abb08afe3f048ea6258b/src/middleware/persist.ts#L9-L20

**Comment by dai-shi** [comment]
> Using the new `createJSONStorage()`, it doesn't seem to be possible anymore to provide a custom `deserialize()` function.

If you want to customize deserialize function, please don't use `createJSONStorage`, but provide your custom storage. That's the design decision we made in #1463.

 See https://github.com/pmndrs/zustand/pull/1763#issuecomment-1518475848 which may cover your use case.

**Reply by dai-shi** [reply, marked as answer]
#1763 is merged and v4.3.8 is released.
